### PR TITLE
RATIS-1106. Add type for DataStream

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Add DataStreamRequestType: HEAD, DATA, CLOSE, START_TRANSACTION

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1106

## How was this patch tested?

New assert

### CODE REVIEW

**Comment by runzhiwang** [comment]
@szetszwo Could you help review this ? Thanks a lot.

File: ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamRequestType.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.protocol;
+
+public enum DataStreamRequestType {
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamRequestType.java:20]: This is indeed a packet type.  Let's change this as an inner class DataStreamPacket.Type.

File: ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamRequestType.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  HEAD,
+  DATA,
+  CLOSE,
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamRequestType.java:23]: Let's call them STREAM_HEADER, STREAM_DATA and STREAM_CLOSE.

**Review by szetszwo — APPROVED** [review]
It probably is a good time to change the packet header to use Protobuf.  Do you want to do it here?  Or we merge this first and do it in a separated JIRA.

The two comments inline are optional since it will be changed again when using Protobuf.

+1

**Comment by runzhiwang** [comment]
@szetszwo Thanks for review. I will use Protobuf in this PR.

**Comment by runzhiwang** [comment]
@szetszwo I have updated the patch. Could you help review it again ?

File: ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamPacketHeader.java [hunk 3]
```diff
@@ -19,29 +19,17 @@
...
-    return SIZE.getSizeInt();
-  }
-
-  public static DataStreamPacketHeader read(LongSupplier readLong, int readableBytes) {
-    if (readableBytes < getSize()) {
-      return null;
-    }
-    return new DataStreamPacketHeader(readLong.getAsLong(), readLong.getAsLong(), readLong.getAsLong());
-  }
+  private static final SizeInBytes SIZE_OF_HEADER_LEN = SizeInBytes.valueOf(4);
```
Comment (amaliujia) [inline ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamPacketHeader.java:27]: why header size is 4 now?

Comment (runzhiwang) [reply ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamPacketHeader.java:27]: It's not header size, it's size of header size. For example, if header size is 24, we only need an int to store 24, and the size of int is 4.

File: ratis-proto/src/main/proto/Raft.proto [hunk 4]
```diff
@@ -279,6 +279,30 @@ message RaftClientRequestProto {
...
+}
+
+message DataStreamRequestHeaderProto {
+  DataStreamPacketHeaderProto packetHeader = 1;
+}
+
+message DataStreamReplyHeaderProto {
+  DataStreamPacketHeaderProto packetHeader = 1;
+  uint64 bytesWritten = 2;
+  bool success = 3;
```
[the remark this answers is not in the document]
Comment (runzhiwang) [reply ratis-proto/src/main/proto/Raft.proto:303]: isSuccess is the name of method.

File: ratis-common/src/main/java/org/apache/ratis/datastream/impl/DataStreamReplyByteBuffer.java [hunk 5]
```diff
@@ -32,15 +33,16 @@
...
+    super(streamId, streamOffset, buffer, type);
 
     this.success = success;
     this.bytesWritten = bytesWritten;
   }
 
   public DataStreamReplyByteBuffer(DataStreamReplyHeader header, ByteBuffer buffer) {
-    this(header.getStreamId(), header.getStreamOffset(), buffer, header.getBytesWritten(), header.isSuccess());
+    this(header.getStreamId(), header.getStreamOffset(), buffer, header.getBytesWritten(), header.isSuccess(),
+        header.getType());
```
Comment (amaliujia) [inline ratis-common/src/main/java/org/apache/ratis/datastream/impl/DataStreamReplyByteBuffer.java:45]: is there a need to include type into replies?

**Review by szetszwo — APPROVED** [review]
+1 the change is perfect!

File: ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamPacketHeader.java (hunk 3 above)
Comment (amaliujia) [reply ratis-common/src/main/java/org/apache/ratis/protocol/DataStreamPacketHeader.java:27]: thanks for the clarification!

**Comment by amaliujia** [comment]
+1

**Comment by runzhiwang** [comment]
@szetszwo Thanks for merging it. @amaliujia Thanks for review.

**Pull request merged**